modules/hardware/raspberryPi/default.nix

```diff
@@ -0,0 +1,270 @@
+    {
+      hardware.socList = [
+        "raspberryPi-arm64"
+      ];
+    }
+
+    (mkIf cfg_rpi.upstreamKernel {
+      hardware.raspberryPi.configTxt.global = {
+        upstream_kernel = true;
+        os_prefix = null;
+        kernel = "Tow-Boot.noenv.bin";
+      };
+    })
+
+    (mkIf cfg.raspberryPi-arm64.enable {
+      system.system = "aarch64-linux";
+
+      hardware.raspberryPi.configTxt = {
+        global.os_prefix = mkDefault "foundation/";
+        filters."all" = mkDefault {};
+      };
+
+      Tow-Boot = {
+        config = [
+          (helpers:
+            with helpers; {
+              # 64 MiB; the default unconfigured state is 4 MiB.
+              SYS_MALLOC_LEN = freeform ''0x4000000'';
+              CMD_POWEROFF = no;
+              CMD_PAUSE = mkForce no;
+
+              # As far as distro_bootcmd is aware, the raspberry pi can
+              # have up to three mmc "devices"
+              #   - https://source.denx.de/u-boot/u-boot/-/blob/v2022.07/include/configs/rpi.h#L134-137
+              # To be fixed in a refresh of the raspberry pi configs.
+              # This currently adds two bogus "SD" entries *sigh*.
+              # It's not an issue upstream since there is no menu; the bootcmd simply tries
+              # all options in order. The bogus entries simply fail.
+              TOW_BOOT_MMC0_NAME = freeform ''"SD (0)"'';
+              TOW_BOOT_MMC1_NAME = freeform ''"SD (1)"'';
+              TOW_BOOT_MMC2_NAME = freeform ''"SD (2)"'';
+            })
+        ];
+
+        patches = [
+          ./0001-rpi_arm64_defconfig-enable-rtl8152.patch
+          ./0001-configs-rpi-allow-for-bigger-kernels.patch
+          ./0001-rpi-Copy-properties-from-firmware-dtb-to-the-loaded-.patch
+        ];
+
+        # The Raspberry Pi firmware expects a filesystem to be used.
+        writeBinaryToFirmwarePartition = false;
+
+        diskImage = {
+          partitioningScheme = "mbr";
+        };
+
+        firmwarePartition = {
+          partitionType = "0C";
+          filesystem = {
+            filesystem = "fat32";
+            populateCommands = ''
+              target="$PWD"
+
+              ## rpi boot config
+              # We assume that a user is customizing config.txt via a custom tow-boot build
+              cp -v "${configTxt}" "$target/config.txt"
+
+              ## rpi firmware / bootloader
+              fwb="${cfg_rpi.firmwarePackage}/share/raspberrypi/boot"
+              cp -vt "$target/" $fwb/bootcode.bin $fwb/fixup*.dat $fwb/start*.elf
+
+              ## rpi firmware DTBs
+
+              ## rpi4 armstubs
+              cp -vt "$target/" "${cfg_rpi.armstubsPackage}/armstub8-gic.bin"
+
+              ## mainline (kernel, dtbs, !overlays)
+              mkdir -p "$target/upstream"
+              cp -vt "$target/upstream/" \
+                "${config.Tow-Boot.outputs.firmware}/binaries/Tow-Boot.noenv.bin" \
+                ${cfg_rpi.mainlineKernel}/dtbs/broadcom/bcm*rpi*.dtb
```

Comment:
Do we really need to compile the whole kernel to generate the dtb? I mean, we definitely could pick a kernel, download all the device trees and compile them ourselves.

Reply:
Sure, I didn't have this problem because my build hit cache (native compile). Just check that you are right and crosscompiling implies build whole kernel.